philander/stc311.py

```diff
@@ -0,0 +1,665 @@
+        Key name                             Value type, meaning and default
+        =================================    ==========================================================================================================
+        TODO: See def_dict
+        ===============================================================================================================================================
+        
+        Also see: :meth:`.Gasgauge.Params_init`, :meth:`.SerialBusDevice.Params_init`, :meth:`.GPIO.Params_init`.
+        """
+        def_dict = {
+            "SerialBusDevice.address": cls.ADDRESSES_ALLOWED[0],
+            "Gasgauge.chip_type": ChipType.STC3115,
+            "Gasgauge.gpio_alarm_idx": None,  # GPIO index for the reset pin
+            "Gasgauge.gpio_cd_idx": None,  # PIO index for the charger driver pin
+            "Gasgauge.senseResistor": 10,  # Sense resistor in milli Ohm
+            "Gasgauge.cc_cnf": 395,  # Coulomb-counter mode configuration
+            "Gasgauge.vm_cnf": 321,  # Voltage mode configuration
+            "Gasgauge.alarm_soc": 2,  # SOC lower threshold; SOC alarm level [0.5%]
+            "Gasgauge.alarm_voltage": 170,  # Voltage lower threshold; 3.0 V
+            "Gasgauge.current_thres": 10,  # Current monitoring threshold; +/-470 V drop
+            "Gasgauge.cmonit_max": 120,  # Monitoring timing threshold; CC-VM: 4 minutes; VM->CC: 1 minute
+            "Gasgauge.pinInt.gpio.Direction": GPIO.DIRECTION_IN  # TODO: these options should be available to be configured and in docs
+        }
+        def_dict.update(paramDict)
+        paramDict.update(def_dict)  # update again to apply changes to original reference
+        return None
+
+    def open(self, paramDict):
+        """Opens the instance and sets it in a usable state.
+
+        Allocate necessary hardware resources and configure
+        user-adjustable parameters to meaningful defaults.
+        In this case the two GPIO-Pins for reading the charger status
+        are initialized.
+        This function must be called prior to any further usage of the
+        instance. Involving it in the system ramp-up procedure could be
+        a good choice. After usage of this instance is finished, the
+        application should call :meth:`close`.
+        
+        :param paramDict(str, object) paramDict: Configuration parameters as obtained from :meth:`Params_init`, possibly.
+        :return: An error code indicating either success or the reason of failure.
+        :rtype: ErrorCode
+        """
+        self.Params_init(paramDict)
+        err = ErrorCode.errOk
+        if paramDict["Gasgauge.chip_type"] == ChipType.STC3115:
+            self.REGISTER = STC3115_Reg
+        elif paramDict["Gasgauge.chip_type"] == ChipType.STC3117:
+            self.REGISTER = STC3117_Reg
+        else:
+            err = ErrorCode.errInvalidParameter
+        self._initialize()
```

Comment:
This function is both used by open and by reset. I could implement it into open, since a reset will require that to be called anyway. As of now, initialize is not implemented completely and not working.